File: ragflow-automation-rearchitecture/selenium-csharp/Base/WebDriverFactory.cs
using AutomationC_.Utils;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Support.UI;

namespace AutomationC_.Base;

public static class WebDriverFactory
{
    private static readonly ThreadLocal<IWebDriver?> DriverHolder = new();
    private static readonly ThreadLocal<WebDriverWait?> WaitHolder = new();

    public static IWebDriver Driver => DriverHolder.Value ?? throw new InvalidOperationException("WebDriver has not been initialized.");

    public static WebDriverWait Wait => WaitHolder.Value ?? throw new InvalidOperationException("WebDriverWait has not been initialized.");

    public static IWebDriver GetDriver(string browser)
    {
        if (DriverHolder.Value != null)
        {
            return DriverHolder.Value;
        }

        IWebDriver driver = browser.ToLowerInvariant() switch
        {
            "chrome" => CreateChromeDriver(),
            "firefox" => CreateFirefoxDriver(),
            _ => throw new ArgumentOutOfRangeException(nameof(browser), browser, "Unsupported browser")
        };

        var implicitWaitInSec = TimeSpan.FromSeconds(
            double.Parse(ConfigReader.GetProperty("IMPLICIT_WAIT_IN_SEC"))
        );
        driver.Manage().Timeouts().ImplicitWait = implicitWaitInSec;

        DriverHolder.Value = driver;
        WaitHolder.Value = new WebDriverWait(driver, implicitWaitInSec);

        return driver;
    }

    public static void CloseBrowser()
    {
        if (DriverHolder.Value is null)
        {
            return;
        }

        try
        {
            DriverHolder.Value.Quit();
        }
        finally
        {
            DriverHolder.Value.Dispose();
            DriverHolder.Value = null;
            WaitHolder.Value = null;
        }
    }

    private static IWebDriver CreateChromeDriver()
    {
        var options = new ChromeOptions();

        if (ConfigReader.GetBooleanProperty("HEADLESS"))
        {
            options.AddArgument("--headless=new");
        }

        if (ConfigReader.GetBooleanProperty("INCOGNITO"))
        {
            options.AddArgument("--incognito");
        }

        options.AddArgument("--disable-notifications");
        options.AddArgument("--disable-gpu");
        options.AddArgument("--start-maximized");

        return new ChromeDriver(options);
    }

    private static IWebDriver CreateFirefoxDriver()
    {
        var options = new FirefoxOptions();

        if (ConfigReader.GetBooleanProperty("HEADLESS"))
        {
            options.AddArgument("--headless");
        }

        if (ConfigReader.GetBooleanProperty("INCOGNITO"))
        {
            options.AddArgument("-private");
        }

        var driverPath = ConfigReader.GetProperty("FIREFOX_DRIVER_PATH");
        if (!string.IsNullOrWhiteSpace(driverPath))
        {
            Environment.SetEnvironmentVariable("webdriver.gecko.driver", driverPath);
        }

        return new FirefoxDriver(options);
    }
}

File: ragflow-automation/src/main/java/org/zemoso/base/WebDriverFactory.java
package org.zemoso.base;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.zemoso.utils.LoggerUtil;

import java.time.Duration;

import static org.zemoso.utils.ConstantUtils.*;

@Slf4j
public class WebDriverFactory {

    private static final ThreadLocal<WebDriverWait> synchronize;
    @Getter
    private static WebDriver driver;

    static {
        synchronize = new ThreadLocal<>();
    }

    private WebDriverFactory() {
    }

    public static WebDriverWait getWait() {
        return synchronize.get();
    }

    public static WebDriver getDriver(String browser) {
        if (driver == null) {
            switch (browser){
                case "chrome":
                    ChromeOptions chromeOptions = new ChromeOptions();
                    if (ConfigReader.getBooleanProperty("HEADLESS")) {
                        LoggerUtil.info("App is running in headless");
                        chromeOptions.addArguments("--headless=new");
                    }

                    // Incognito
                    if (ConfigReader.getBooleanProperty("INCOGNITO")) {
                        chromeOptions.addArguments("--incognito");
                    }

                    chromeOptions.addArguments("--disable-notifications");
                    chromeOptions.addArguments("--disable-gpu");
                    chromeOptions.addArguments("--start-maximized");

                    driver = new ChromeDriver(chromeOptions);
                    break;
                case "firefox":
                    System.setProperty("webdriver.gecko.driver", ConfigReader.getProperty("FIREFOX_DRIVER_PATH"));

                    FirefoxOptions firefoxOptions = new FirefoxOptions();

                    // Headless
                    if (ConfigReader.getBooleanProperty("HEADLESS")) {
                        log.info("in head less");
                        firefoxOptions.addArguments("--headless");
                    }

                    // Private browsing (incognito)
                    if (ConfigReader.getBooleanProperty("INCOGNITO")) {
                        firefoxOptions.addArguments("-private");
                    }

                    driver = new FirefoxDriver(firefoxOptions);
                    break;
                default:
                    throw new RuntimeException("Unsupported browser: " + browser);
            }

            long implicitWait = Long.parseLong(ConfigReader.getProperty("IMPLICIT_WAIT_IN_SEC"));
            driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(implicitWait));

        }

        return driver;
    }

    public static void closeBrowser() {
        if (driver != null) {
            driver.quit();
        }
    }
}
